Design-Docs/standardizing-cluster-infrastructure.md

```diff
@@ -0,0 +1,277 @@
+- Easy certification process for cloud providers
+- Concise documentation on how to fulfill SCS certification
+- Applications are configured independently of the cloud provider
+- Automated e2e tests for cluster and application configuration
+- Secure cluster infrastructure for users and cluster operators
+- Auditability & automated testing (unit and e2e tests) of SCS code
+
+### Non-goals
+
+- This proposal does not intend to define specifically all inputs, e.g. to define load balancers in SCS infrastructure. This is left for an open discussion in the SCS team and with cloud providers.
+- We don’t want to say that any specific components are must-haves for cluster infrastructure, but we will give suggestions.
+
+## Proposal
+
+This proposal does not specify any details about load balancer definitions, node images, or server types. It rather aims to define a way to handle standardization and certification in general.
+
+It consists of three parts:
+
+### Proposal 1: Pre-defined cluster stacks are necessary to ensure that clusters and applications work - a free configuration of cluster components is not possible
+
+A cluster stack consists of “big components”, e.g. node image and Cluster API provider; down to “small” ones, such as private/public network, IPv4/IPv6, container runtime (interface), and firewall. All of these components related to the image, networking, or cluster resources interact and there is no way of freely configuring such a cluster without heavy constraints.
+
+Why do we think that such cluster stacks are necessary?  
+
+- Some configurations are incompatible (e.g. choosing eBPF without a node image that supports it)
+- Some configurations might work but require further knowledge and additional work (Ubuntu and Cilium work together in general, but not with Cilium’s latest version because of a bug in older kernels)
+- Some configurations depend on the cloud providers. A user has to know these details to get a working system (MTU size, e.g. with Cilium).
+
+Assuming that a cluster consists of n components, then we would say that decisions need to be taken for all of these n components (i.e. Kubeadm as bootstrap controller or Ubuntu as OS for node image). The collection of these decisions defines a cluster stack.
+
+There can be multiple cluster stacks to satisfy different needs, but there should not be too many cluster stacks. This is necessary to keep the cluster stacks maintainable (defining, testing, updating) and to not overwhelm and confuse users with a large number of stacks that cannot be clearly differentiated.
+
+### Proposal 2: Cluster operators should be able to define clusters independently of the provider
+
+The cluster infrastructure should be independent of the providers because of a few reasons. It ensures true freedom of choice (a big aspect of digital sovereignty), allows easy migrations of infrastructure from one provider to another, and keeps the configuration of clusters easy for cluster operators.
```

Comment:
"allows easy migrations of infrastructure from one provider to another" is probably not true. At most, it simplifies migrations. But as soon as you want to reduce downtimes of your services, no migration between two clusters is ever "easy".

Reply:
I think a distinction should be made here between the basic interchangeability of infrastructures and a migration. I think it's not necessarily the first goal to achieve migration between different infrastructures.

Reply:
@horazont @berendt we certainly meant inter-changeability, which is a necessary part of migrating clusters from one provider to another easily. However, there are many more things that play a role. Therefore, we mean "inter-changeability" with migrations in the text.